{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8-final"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3",
   "language": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "source": [
    "# Process Bacteria\n",
    "\n",
    "... description of the notebook goes here ...\n",
    "\n",
    "### Prerequisites:\n",
    "- conda (Miniconda or Anaconda): https://conda.io/projects/conda/en/latest/user-guide/install/index.html\n",
    "- scikit-bio: `conda install -c https://conda.anaconda.org/biocore scikit-bio`\n",
    "- prophage tools & their setup (follow instructions in section 2.)"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "source": [
    "## 1. Notebook Setup\n",
    "\n",
    "Import the necessary libraries and set the data & output paths"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "\n",
    "import skbio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to the folder containing genomic data (relative to current working directory)\n",
    "data_dir = Path(Path.cwd(), '..', 'data')\n",
    "data_dir.mkdir(exist_ok=True)\n",
    "(data_dir / 'compressed').mkdir(exist_ok=True)\n",
    "\n",
    "# Path to the output folder\n",
    "out_dir = Path(Path.cwd(), '..', 'out')\n",
    "out_dir.mkdir(exist_ok=True)"
   ]
  },
  {
   "source": [
    "## 2. Prophage Tool Setup\n",
    "\n",
    "Select prophage CLI tools you intend to use. The recommended ones include:\n",
    "\n",
    " - PhageBoost: https://github.com/ku-cbd/PhageBoost\n",
    " - PhiSpy: https://github.com/linsalrob/PhiSpy\n",
    "\n",
    "For each tool:\n",
    "\n",
    " 1. Create new conda enviroment: `conda create -n [name_of_the_tool]`\n",
    " 1. Install the tool and its prerequisites according to author's instructions\n",
    " 1. Register the path to the binary and arguments the tool should be invoked with using the function below:"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "tools = []\n",
    "\n",
    "def register_tool(tool_name, tool_path, input_arg, output_arg, tool_input_type='fasta', separate_subsequences=False, **cli_args):\n",
    "    def invoke_tool(inp, out):\n",
    "        return f'{tool_path} {input_arg} {inp} {output_arg} {out} {\" \".join([f\"--{f} {a}\" for f, a in cli_args.items()])}'\n",
    "    tools.append({\n",
    "        'name': tool_name,\n",
    "        'input_type': tool_input_type,\n",
    "        'separate_subsequences': separate_subsequences,\n",
    "        'cli_call': invoke_tool\n",
    "    })\n",
    "\n",
    "# Register tools here\n",
    "register_tool('PhageBoost',\n",
    "    tool_path='~/miniconda3/envs/phage-boost/bin/PhageBoost',\n",
    "    input_arg='-f',\n",
    "    output_arg='-o',\n",
    "    # Other args go here\n",
    ")\n",
    "\n",
    "register_tool('PhiSpy',\n",
    "    tool_input_type='gb',\n",
    "    tool_path='~/miniconda3/envs/phispy/bin/PhiSpy.py',\n",
    "    input_arg='',\n",
    "    output_arg='-o',\n",
    "    # Other args go here\n",
    ")"
   ]
  },
  {
   "source": [
    "## 3. Data extraction\n",
    "\n",
    "Download genomic data for the bacteria of interest from [NCBI Assembly](https://www.ncbi.nlm.nih.gov/assembly) in fasta (.fna) and genbank (.gbff) formats. Name them \"[taxid]_(fasta|gb).tar\" and put them in \"compressed\" folder in your data folder, then run the following code:"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change to True to extract data anew\n",
    "overwrite_data = False\n",
    "\n",
    "# 1.c decompress all downloaded assemblies\n",
    "for entry in os.scandir(data_dir / 'compressed'):\n",
    "    if entry.is_file():\n",
    "        taxid, file_type = entry.name.split('.')[0].split('_')\n",
    "        file_path = data_dir / taxid / file_type\n",
    "\n",
    "        if file_path.exists():\n",
    "            if overwrite_data:\n",
    "                shutil.rmtree(file_path)\n",
    "            else:\n",
    "                continue\n",
    "\n",
    "        file_path.mkdir(exist_ok=True, parents=True)\n",
    "\n",
    "        # Run tar command in a subprocess to extract the archive\n",
    "        subprocess.run(['tar', '-xf', entry.path, '-C', file_path, '--strip-components', '1'])\n",
    "\n",
    "        # Unzip all extracted files\n",
    "        for f in os.scandir(file_path):\n",
    "            if (f.path.endswith('.gz')):\n",
    "               subprocess.run(['gzip', '-d', f.path])\n",
    "                # Rename the files to include only assembly accession id\n",
    "               f_path = Path(f.path.replace('.gz', ''))\n",
    "               f_path.rename(f_path.parents[0] / ('_'.join(f_path.stem.split('_')[:2]) + f_path.suffix))\n",
    "            else:\n",
    "                # Remove files that don't contain genomic sequences\n",
    "                os.remove(f.path)"
   ]
  },
  {
   "source": [
    "### 3.1 Flatten FASTA files\n",
    "\n",
    "Many tools work better with only one sequence per fasta file, so we need to split the files with multiple sequences into individual ones:"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "for entry in os.scandir(data_dir):\n",
    "    fasta_dir = Path(entry.path, 'fasta')\n",
    "    if entry.is_dir() and fasta_dir.exists():\n",
    "        for f in os.scandir(fasta_dir):\n",
    "            if not f.is_file():\n",
    "                continue\n",
    "            \n",
    "            f_path = Path(f.path)\n",
    "            dir_path = (f_path.parents[0] / f_path.stem)\n",
    "            dir_path.mkdir(exist_ok=True)\n",
    "\n",
    "            for seq in skbio.io.read(str(f_path), format='fasta'):\n",
    "                seq.write(str(dir_path / (seq.metadata['id'] + '.fasta')), format='fasta')\n",
    "            \n",
    "            os.remove(f_path)"
   ]
  },
  {
   "source": [
    "## 4. Run the tools\n",
    "\n",
    "Predict the prophages from the data sequences and save the results in \\[output_folder\\]/\\[tool_name\\]/\\[taxid\\]/\\[assembly_accession\\]/:"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "ing_model.html\n",
      "time after predictions: 2.2317392826080322\n",
      "{}\n",
      "no phages found \n",
      "processing: WOUN01000007\n",
      "time after genecalls: 0.13173556327819824\n",
      "time after feature calculations: 0.4884936809539795\n",
      "[11:51:49] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 2.0665030479431152\n",
      "{}\n",
      "no phages found \n",
      "processing: WOUN01000001\n",
      "time after genecalls: 1.5512664318084717\n",
      "time after feature calculations: 4.097052574157715\n",
      "[11:51:56] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 5.723111152648926\n",
      "{}\n",
      "no phages found \n",
      "processing: WOUN01000002\n",
      "time after genecalls: 1.9734890460968018\n",
      "time after feature calculations: 5.288930416107178\n",
      "[11:52:04] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 6.832658052444458\n",
      "{'phage0': 12}\n",
      "processing: WOUN01000003\n",
      "time after genecalls: 0.2137138843536377\n",
      "time after feature calculations: 0.6973423957824707\n",
      "[11:52:07] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 2.1347999572753906\n",
      "{}\n",
      "no phages found \n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "/home/chris/prophage-tool/notebooks/../data/556287/fasta/GCA_001414235.1\n",
      "\n",
      "\n",
      "processing: LLVZ01000025\n",
      "after min_size_of_contig 10000, no contigs were left. This caused the run to fail\n",
      "processing: LLVZ01000015\n",
      "after min_size_of_contig 10000, no contigs were left. This caused the run to fail\n",
      "processing: LLVZ01000014\n",
      "after min_size_of_contig 10000, no contigs were left. This caused the run to fail\n",
      "processing: LLVZ01000020\n",
      "after min_size_of_contig 10000, no contigs were left. This caused the run to fail\n",
      "processing: LLVZ01000013\n",
      "after min_size_of_contig 10000, no contigs were left. This caused the run to fail\n",
      "processing: LLVZ01000011\n",
      "after min_size_of_contig 10000, no contigs were left. This caused the run to fail\n",
      "processing: LLVZ01000018\n",
      "after min_size_of_contig 10000, no contigs were left. This caused the run to fail\n",
      "processing: LLVZ01000012\n",
      "after min_size_of_contig 10000, no contigs were left. This caused the run to fail\n",
      "processing: LLVZ01000021\n",
      "after min_size_of_contig 10000, no contigs were left. This caused the run to fail\n",
      "processing: LLVZ01000002\n",
      "time after genecalls: 1.0035679340362549\n",
      "time after feature calculations: 2.8746633529663086\n",
      "[11:52:23] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 4.5312535762786865\n",
      "{}\n",
      "no phages found \n",
      "processing: LLVZ01000017\n",
      "after min_size_of_contig 10000, no contigs were left. This caused the run to fail\n",
      "processing: LLVZ01000024\n",
      "after min_size_of_contig 10000, no contigs were left. This caused the run to fail\n",
      "processing: LLVZ01000023\n",
      "after min_size_of_contig 10000, no contigs were left. This caused the run to fail\n",
      "processing: LLVZ01000005\n",
      "time after genecalls: 0.3006019592285156\n",
      "time after feature calculations: 0.9276340007781982\n",
      "[11:52:30] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 2.4733359813690186\n",
      "{}\n",
      "no phages found \n",
      "processing: LLVZ01000003\n",
      "time after genecalls: 0.667978048324585\n",
      "time after feature calculations: 1.8323523998260498\n",
      "[11:52:35] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 3.400373697280884\n",
      "{'phage1': 16, 'phage2': 12}\n",
      "processing: LLVZ01000026\n",
      "after min_size_of_contig 10000, no contigs were left. This caused the run to fail\n",
      "processing: LLVZ01000009\n",
      "after min_size_of_contig 10000, no contigs were left. This caused the run to fail\n",
      "processing: LLVZ01000019\n",
      "after min_size_of_contig 10000, no contigs were left. This caused the run to fail\n",
      "processing: LLVZ01000016\n",
      "after min_size_of_contig 10000, no contigs were left. This caused the run to fail\n",
      "processing: LLVZ01000007\n",
      "time after genecalls: 0.18677854537963867\n",
      "time after feature calculations: 0.640134334564209\n",
      "[11:52:43] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 2.2440340518951416\n",
      "{}\n",
      "no phages found \n",
      "processing: LLVZ01000010\n",
      "after min_size_of_contig 10000, no contigs were left. This caused the run to fail\n",
      "processing: LLVZ01000022\n",
      "after min_size_of_contig 10000, no contigs were left. This caused the run to fail\n",
      "processing: LLVZ01000004\n",
      "time after genecalls: 0.4871096611022949\n",
      "time after feature calculations: 1.4051463603973389\n",
      "[11:52:50] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 3.055051803588867\n",
      "{}\n",
      "no phages found \n",
      "processing: LLVZ01000008\n",
      "time after genecalls: 0.127166748046875\n",
      "time after feature calculations: 0.4412209987640381\n",
      "[11:52:53] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 2.0777218341827393\n",
      "{}\n",
      "no phages found \n",
      "processing: LLVZ01000006\n",
      "time after genecalls: 0.17569613456726074\n",
      "time after feature calculations: 0.6827223300933838\n",
      "[11:52:56] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 2.2997758388519287\n",
      "{}\n",
      "no phages found \n",
      "processing: LLVZ01000001\n",
      "time after genecalls: 1.0198509693145752\n",
      "time after feature calculations: 3.0593671798706055\n",
      "[11:53:02] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 4.5879902839660645\n",
      "{}\n",
      "no phages found \n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "/home/chris/prophage-tool/notebooks/../data/28105/fasta/GCA_002197105.1\n",
      "\n",
      "\n",
      "processing: CP021826\n",
      "time after genecalls: 0.7701220512390137\n",
      "time after feature calculations: 2.0000319480895996\n",
      "[11:53:07] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 3.615894079208374\n",
      "{}\n",
      "no phages found \n",
      "processing: CP021828\n",
      "time after genecalls: 7.051033020019531\n",
      "time after feature calculations: 17.114713668823242\n",
      "[11:53:27] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 18.729459762573242\n",
      "{}\n",
      "no phages found \n",
      "processing: CP021827\n",
      "time after genecalls: 5.663811922073364\n",
      "time after feature calculations: 13.841742038726807\n",
      "[11:53:43] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 15.548301219940186\n",
      "{}\n",
      "no phages found \n",
      "processing: CP021825\n",
      "time after genecalls: 15.741645097732544\n",
      "time after feature calculations: 38.31173658370972\n",
      "[11:54:24] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 40.0517942905426\n",
      "{'phage1': 11}\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "/home/chris/prophage-tool/notebooks/../data/1273132/fasta/GCA_000325745.1\n",
      "\n",
      "\n",
      "processing: CP003789\n",
      "time after genecalls: 4.89551043510437\n",
      "time after feature calculations: 14.210298299789429\n",
      "[11:54:42] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 15.830983638763428\n",
      "{'phage0': 27, 'phage3': 20}\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "/home/chris/prophage-tool/notebooks/../data/234/fasta/GCA_014884505.1\n",
      "\n",
      "\n",
      "processing: CP054955\n",
      "time after genecalls: 9.089780807495117\n",
      "time after feature calculations: 20.997640132904053\n",
      "[11:55:06] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 22.662135124206543\n",
      "{'phage0': 21}\n",
      "processing: CP054956\n",
      "time after genecalls: 6.388880014419556\n",
      "time after feature calculations: 15.030055046081543\n",
      "[11:55:23] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 16.706812858581543\n",
      "{}\n",
      "no phages found \n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "/home/chris/prophage-tool/notebooks/../data/34020/fasta/GCA_001021085.1\n",
      "\n",
      "\n",
      "processing: CP004021\n",
      "time after genecalls: 3.869270086288452\n",
      "time after feature calculations: 11.029396295547485\n",
      "[11:55:37] WARNING: ../src/tree/./updater_quantile_hist.h:135: Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost. A likely cause for this warning is that the model was saved with saveRDS() in R or pickle.dump() in Python. We strongly ADVISE AGAINST using saveRDS() or pickle.dump() so that the model remains accessible in current and upcoming XGBoost releases. Please use xgb.save() instead to preserve models for the long term. For more details and explanation, see https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html\n",
      "time after predictions: 12.729522228240967\n",
      "{'phage1': 28, 'phage2': 22}\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "/home/chris/prophage-tool/notebooks/../data/773/gb/GCA_002007505.1.gbff\n",
      "\n",
      "\n",
      "Making Test Set...\n",
      "Start Classification Algorithm...\n",
      "As training flag is zero, considering unknown functions\n",
      "Evaluating...\n",
      "Checking prophages in initial_tbl.tsv\n",
      "\n",
      "Potential prophages (sorted highest to lowest)\n",
      "Contig\tStart\tStop\tNumber of potential genes\tStatus\n",
      "CP015821.1\t594334\t639376\t48\tKept\n",
      "CP015821.1\t1563440\t1602516\t46\tKept\n",
      "CP015821.1\t1993944\t2031362\t36\tKept\n",
      "CP015821.1\t2324391\t2351733\t12\tKept\n",
      "CP015821.1\t1236059\t1243914\t5\tKept\n",
      "CP015821.1\t2578155\t2582098\t2\tDropped. Not enough genes\n",
      "CP015821.1\t32943\t33881\t1\tDropped. Not enough genes\n",
      "CP015821.1\t1297830\t1298138\t1\tDropped. Not enough genes\n",
      "CP015821.1\t2267984\t2268391\t1\tDropped. Not enough genes\n",
      "CP015821.1\t2512370\t2513323\t1\tDropped. Not enough genes\n",
      "\n",
      "PROPHAGE: 1 Contig: CP015821.1 Start: 594334 Stop: 639376\n",
      "PP: 1 start: 594334 stop: 639376\n",
      "Finding repeats in pp 1 from 592334 to 641376\n",
      "There were 3 repeats with the same length as the best. One chosen somewhat randomly!\n",
      "PROPHAGE: 2 Contig: CP015821.1 Start: 1236059 Stop: 1243914\n",
      "PP: 2 start: 1236059 stop: 1243914\n",
      "Finding repeats in pp 2 from 1234059 to 1245914\n",
      "PROPHAGE: 3 Contig: CP015821.1 Start: 1563440 Stop: 1602516\n",
      "PP: 3 start: 1563440 stop: 1602516\n",
      "Finding repeats in pp 3 from 1561440 to 1604516\n",
      "There were 9 repeats with the same length as the best. One chosen somewhat randomly!\n",
      "PROPHAGE: 4 Contig: CP015821.1 Start: 1993944 Stop: 2031362\n",
      "PP: 4 start: 1993944 stop: 2031362\n",
      "Finding repeats in pp 4 from 1991944 to 2033362\n",
      "There were 25 repeats with the same length as the best. One chosen somewhat randomly!\n",
      "PROPHAGE: 5 Contig: CP015821.1 Start: 2324391 Stop: 2351733\n",
      "PP: 5 start: 2324391 stop: 2351733\n",
      "Finding repeats in pp 5 from 2322391 to 2353733\n",
      "Done!!!\n",
      "\n",
      "--------------------------------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Change to True to process data anew\n",
    "overwrite_output = True\n",
    "\n",
    "for tool in tools:\n",
    "    for entry in os.scandir(data_dir):\n",
    "        # Use only data in the correct format for the tool\n",
    "        seq_dir = Path(entry.path, tool['input_type'])\n",
    "        if entry.is_dir() and seq_dir.exists():\n",
    "            for assembly in os.scandir(seq_dir):\n",
    "                print(assembly.path)\n",
    "                print('\\n')\n",
    "\n",
    "                # Create output path\n",
    "                tool_out_path = out_dir / tool['name'] / entry.name / '.'.join(assembly.name.split('.')[:2])\n",
    "                if tool_out_path.exists():\n",
    "                    if overwrite_output or not (tool_out_path / '.done').exists():\n",
    "                        shutil.rmtree(tool_out_path)\n",
    "                    else:\n",
    "                        print('\\n--------------------------------------------------\\n')\n",
    "                        continue\n",
    "\n",
    "                tool_out_path.mkdir(exist_ok=True, parents=True)\n",
    "\n",
    "                # Run the tool\n",
    "                if assembly.is_file():\n",
    "                    !{tool['cli_call'](assembly.path, str(tool_out_path))}\n",
    "                elif assembly.is_dir():\n",
    "                    for f in os.scandir(assembly.path):\n",
    "                        !{tool['cli_call'](f.path, str(tool_out_path / '.'.join(f.name.split('.')[:-1])) if tool['separate_subsequences'] else str(tool_out_path))}\n",
    "\n",
    "                # Mark as done\n",
    "                (tool_out_path / '.done').touch()\n",
    "                print('\\n--------------------------------------------------\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ]
}